### ISSUE TITLE
CDN with FTPS (proftpd @ Ubuntu 20.04): Uploads work, but "Test FTP server" fails

### ISSUE THREAD

**Issue by caugner** [W3EDGE/w3-total-cache#310]
I have a WordPress single instance configured with W3 Total Cache that uses an Ubuntu 20.04 Server with proftpd + nginx for the CDN. Even though uploads work without problems, I'm getting the following error when clicking on "Test FTP server":

```
Error: Unable to upload file: /www/htdocs/***/www/wp-content/cache/tmp/test_file_88406f892caee8cec1cb7c743676ea0e (ftp_put(): Unable to build data connection: Connection refused).
```

I checked the CDN manually with FileZilla and found a directory `test_dir_88406f892caee8cec1cb7c743676ea0e` with an empty file `test_file_88406f892caee8cec1cb7c743676ea0e` inside the configured "FTP path". What could be the problem?

**Comment by mavas84** [comment]
Hello @caugner 

Thank you for taking the time to report this.
Just to confirm, under Performance>General settings>CDN, you selected Origin push Self-hosted / File Transfer Protocol Upload?
Have you enabled Use passive FTP mode? Can you also please share the screenshot of the settings in Performance>CDN so we can try and replicate this?
Thanks!

**Comment by mavas84** [comment]
Closing due to inactivity.

**Comment by caugner** [comment]
@mavas84 I would like to reopen this topic, because we're still affected by this issue, and now uploads are failing as well.

We're using a Ubuntu 20.04 LTS server in the Hetzner Cloud with proftpd installed. W3 Total Cache is configured as follows (CDN section):

- [x] Use passive FTP mode
- FTP hostname: `cdn.***.de`
- FTP-connection: `SSL-FTP connection (FTPS)`
- FTP username: `***`
- FTP password: `***`
- SSL support: `Enabled (always use SSL)`
- [x] Use default SSH public/private key files
- SFTP public key: (empty)
- SFTP private key: (empty)
- Replace site's hostname with: `cdn.***.de`

When W3 Total Cache connects to the server, I can see the following log entries in the `/var/log/proftpd/tls.log` (and there is no sign of the connection being refused):

```
2021-06-28 12:42:43,292 mod_tls/2.7[1105]: TLS/TLS-C requested, starting TLS handshake
2021-06-28 12:42:43,342 mod_tls/2.7[1105]: client supports secure renegotiations
2021-06-28 12:42:43,342 mod_tls/2.7[1105]: TLSv1.3 connection accepted, using cipher TLS_AES_256_GCM_SHA384 (256 bits)
2021-06-28 12:42:43,389 mod_tls/2.7[1105]: Protection set to Private
```

**Comment by caugner** [comment]
I found a more meaningful error message by proftpd now:

```
Error connecting to **.**.**.**#37941 for active data transfer: Connection refused
```

This sounds like W3 Total Cache is not initiating a **passive** SFTP connection, but an active one.

**Comment by caugner** [comment]
@mavas84 I just noticed in the Browser developer tools' "Network" panel that the request triggered by the "Test FTP server" always sends `config[pasv]: "0"`, even if the "Use passive FTP mode" is checked. So that's most likely what causes the error.

**Comment by caugner** [comment]
@mavas84 Found the culprit, see: #393

**Comment by mavas84** [comment]
Hello @caugner 

Thank you for the details.
The devs will make sure to test the pull request and get back to you as soon as the tests are complete.
Thanks!

**Comment by mavas84** [comment]
Hello @caugner 

Thank you once again for taking the time to report this and especially for the pull request. This is now fixed in the latest 2.1.6 release!
